### PR TITLE
Add win_unicode_console dependency

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Adds `win_unicode_console` dependency for Py2 on Windows
Enables `win_unicode_console` in the `sitecustomization.py` file

https://pypi.python.org/pypi/win_unicode_console

This needs to be discussed
@cachedout ^^^

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/36011

### Previous Behavior
In Windows with Python 2 parameters passed on the command line were not passed properly as Unicode.

### New Behavior
Unicode characters are now passed properly as Unicode.

### Tests written?
NA

### CODE REVIEW

**Comment by cachedout** [comment]
I don't see where this does anything but add the dependency. How does this integrate with Salt?

**Comment by twangboy** [comment]
@cachedout It creates a `sitecustomize.py` file in the `site-packages` directory which gets loaded when python starts. The contents of the `sitecustomize.py` file:

```
import win_unicode_console
win_unicode_console.enable(use_unicode_argv=True)
```

**Comment by cachedout** [comment]
Wait, in the `site-packages` directory? So this change affects all python libs and not just salt?

**Comment by twangboy** [comment]
@cachedout Yes

**Comment by skizunov** [comment]
It shouldn't be an issue since on Windows, Python and associated libraries are private to the Salt installation. I do question the need for Python 2 support though. On Windows Salt-Master under Python 2, the TCP load balancer is disabled (due to missing functionality in Python 2 for pickling sockets). It cripples the Windows Salt-Master when using the TCP transport. Since Python 3 works just fine and the version of Python is private to Salt, I don't see the need to maintain and support both flavors at the installer level.

**Comment by twangboy** [comment]
A vote for moving completely to Py3 in Windows....

**Comment by damon-atkins** [comment]
As I commented in other PR etc....   Yep PY2 needs to be force into supporting new versions of Windows. PY2 was ok for WinXP.  2017.5.X should be last PY2 release on Windows.   Move forward with 3.6.1 which has the console Unicode fixes.

**Comment by skizunov** [comment]
@damon-atkins : I would actually go one step further than what you are suggesting and not even have a Windows installer for PY2 on 2017.5.x. The reason is that unlike Linux, the Windows installer comes with a separate, self-contained environment that is completely independent from any python version(s) that are installed on the system. So if the user wanted to use PY2 for other tasks, they are free to do so and it won't interfere with their Salt installation. Giving a choice between PY2 and PY3 installers to the end-user may serve to confuse them because they would have to think about which one to choose. I would say just give them the one option that provides the most functionality and relieve the end-user of the burden of making a choice.

To balance my perspective, one reason why PY2 may be attractive to some users is if they have their own salt plugins that don't work on PY3 or if they have a `#!py` based SLS file which is not PY3 compatible. But I don't think this is a common case for Windows users.

**Comment by thatch45** [comment]
Yes, I want to do both because I don't want to have unexpected issues that turn into hard breaking issues pop up instead of being able to more easily roll fixes into the python3 version.
So the py2 will be the default download and then we will make the py3 the only download after it has been out for a while, likely the release after Nitrogen

**Comment by skizunov** [comment]
@thatch45 : Sounds fair. A related question, for the Windows PY3 variant, why did you choose to go with Python 3.5.3 and not Python 3.6.1? Is it due to being forced to move to pycryptdome when using Python 3.6.1 (due to the unavailability of pycrypto for that version)? Or are there other dependencies or concerns with moving to it? I think Python 3.6.1 is attractive simply due to the speed and memory usage optimizations it has made over 3.5.x (although I have no hard benchmarks yet, so it is theoretical at this point). @damon-atkins seems to like it for other reasons (eg Windows Console Unicode fixes).

**Comment by thatch45** [comment]
It was really based on the maturity of the python branch, that 3.5 has been out longer. I want to move to 3.6 on the release after Nitrogen.

**Comment by damon-atkins** [comment]
@skizunov    You cannot take the PY2 version away from clients without giving them notice. Best bet is to drop PY2 support with the first 2018 release with warnings saying the PY3 is preview state for 2017.

@thatch45 Could the salt team on on windows only
PY2 on startup warning "Python 2 version of Salt will be deprecated with the first release of 2018 version"
PY3 on startup warning "Python 3 version of Salt is a preview release"

PY 3.6.2 will be out the middle of this year.

**Comment by thatch45** [comment]
@damon-atkins that would be a wise thing to do.
@twangboy lets discuss this, it would not be a blocker for the Nitrogen release and would be ok to get into a point release

**Review by thatch45 — APPROVED** [review]

**Comment by thatch45** [comment]
I talked this over with @twangboy and I think that it is the right move.

**Comment by damon-atkins** [comment]
Given this is merged, might be worth changing the title.

**Comment by cachedout** [comment]
@damon-atkins Do you have a suggested title?

**Comment by twangboy** [comment]
@cachedout I changed the title... remove the `[DO NOT MERGE]`

**Comment by damon-atkins** [comment]
Just thought about this.
How is 'mbcs'  use impacted by this change of changing the cp e.g.  instr.encode('mbcs')  ?
i.e. does changing the code page for salt-call change 'mbcs'  does it also mean salt-minion daemon will run differently. 

I suppose it comes down to were 'mbcs'  gets its vaule from the current running shell or an overall OS value. 

>on Windows, Python uses the name “mbcs” to refer to whatever the currently configured encoding is

**Comment by twangboy** [comment]
I'm running into problems with the `win_unicode_console` stuff in my dev environment when installing additional packages using pip... some packages fail to install. We may need to rethink this. Maybe just bag this completely and say that if you need full unicode support to use Python3...

**Comment by damon-atkins** [comment]
Keep in mine that Py 3.6.1 has the console unicode fix, until Windows is on this people will still have issues.  (3.6.2 out end of June).   PY2 might be in the two hard basket to aim for perfect support.   Add the warning that PY2 will be dropped on windows in 2018 first release of salt.

**Comment by damon-atkins** [comment]
re-awake the discussion on the version of python in Oxygen i.e. Python 3.6.3+ to fix Unicode issues.
And add a warning that this is the last Python 2 release.

**Comment by twangboy** [comment]
@damon-atkins I don't think we're planning on moving to 3.6 until next major release (Fluorine). We need to make sure all the deps work with it.

**Comment by damon-atkins** [comment]
I would have thought with @terminalmage effort and others to get Unicode across the line, that that these two fixes would help.
PEP 528, Change Windows console encoding to UTF-8
PEP 529, Change Windows filesystem encoding to UTF-8
(PY2 still has these and other issues)

Changes on Windows in 3.6.2
bpo-30418: On Windows, subprocess.Popen.communicate() now also ignore EINVAL on stdin.write() if the child process is still running but closed the pipe.
Changes on Windows in 3.6.0
bpo-28208: Update Windows build and OS X installers to use SQLite 3.14.2.
bpo-28248: Update Windows build and OS X installers to use OpenSSL 1.0.2j.
bpo-28200: Fix memory leak on Windows in the os module (fix path_converter() function).
bpo-27932: Prevent memory leak in win32_ver().
bpo-28402: Adds signed catalog files for stdlib on Windows.
bpo-28333: Enables Unicode for ps1/ps2 and input() prompts. (Patch by Eryk Sun)
bpo-28251: Improvements to help manuals on Windows.
bpo-28110: launcher.msi has different product codes between 32-bit and 64-bit
bpo-28161: Opening CON for write access fails
bpo-28162: WindowsConsoleIO readall() fails if first line starts with Ctrl+Z
bpo-28163: WindowsConsoleIO fileno() passes wrong flags to _open_osfhandle
bpo-28164: _PyIO_get_console_type fails for various paths
bpo-28137: Renames Windows path file to ._pth
bpo-28138: Windows ._pth file should allow import site
bpo-17909: json.load and json.loads now support binary input encoded as UTF-8, UTF-16 or UTF-32. Patch by Serhiy Storchaka.
bpo-28065: Update xz dependency to 5.2.2 and build it from source.
bpo-25144: Ensures TargetDir is set before continuing with custom install.
bpo-1602: Windows console doesn’t input or print Unicode (PEP 528)
bpo-27781: Change file system encoding on Windows to UTF-8 (PEP 529)
bpo-27731: Opt-out of MAX_PATH on Windows 10
bpo-6135: Adds encoding and errors parameters to subprocess.
bpo-27959: Adds oem encoding, alias ansi to mbcs, move aliasmbcs to codec lookup.
bpo-27982: The functions of the winsound module now accept keyword arguments.
bpo-20366: Build full text search support into SQLite on Windows.
bpo-27756: Adds new icons for Python files and processes on Windows. Designs by Cherry Wang.
bpo-27883: Update sqlite to 3.14.1.0 on Windows.
bpo-26930: Update Windows builds to use OpenSSL 1.0.2h.
bpo-27064: The py.exe launcher now defaults to Python 3. The Windows launcher py.exe no longer prefers an installed Python 2 version over Python 3 by default when used interactively.
bpo-27053: Updates make_zip.py to correctly generate library ZIP file.
bpo-26268: Update the prepare_ssl.py script to handle OpenSSL releases that don’t include the contents of the include directory (that is, 1.0.2e and later).
bpo-26071: bdist_wininst created binaries fail to start and find 32bit Python
bpo-26073: Update the list of magic numbers in launcher
bpo-26065: Excludes venv from library when generating embeddable distro.
bpo-25022: Removed very outdated PC/example_nt/ directory.
Changes in 3.5.3
bpo-28251: Improvements to help manuals on Windows.
bpo-28110: launcher.msi has different product codes between 32-bit and 64-bit
bpo-25144: Ensures TargetDir is set before continuing with custom install.
bpo-27469: Adds a shell extension to the launcher so that drag and drop works correctly.
bpo-27309: Enabled proper Windows styles in python[w].exe manifest.
bpo-25339: PYTHONIOENCODING now has priority over locale in setting the error handler for stdin and stdout.
bpo-27114: Fix SSLContext._load_windows_store_certs fails with PermissionError
bpo-27053: Updates make_zip.py to correctly generate library ZIP file.
bpo-26268: Update the prepare_ssl.py script to handle OpenSSL releases that don’t include the contents of the include directory (that is, 1.0.2e and later).
bpo-26071: bdist_wininst created binaries fail to start and find 32bit Python
bpo-26073: Update the list of magic numbers in launcher
bpo-26065: Excludes venv from library when generating embeddable distro.

**Pull request merged**
